test/unit/vpc.test.ts

```diff
@@ -0,0 +1,75 @@
+import { get } from "lodash";
+import { merge } from "lodash";
+import { baseConfig, pluginConfigExt, runServerless } from "../utils/runServerless";
+import ServerlessError from "../../src/utils/error";
+import {AwsCfInstruction} from "@serverless/typescript";
+
+describe("vpc", () => {
+    it("should have subnets and security groups", async () => {
+        const { cfTemplate, computeLogicalId } = await runServerless({
+            fixture: "vpc",
+            configExt: pluginConfigExt,
+            command: "package",
+        });
+
+        const vpcConfig = get(cfTemplate.Resources.FooLambdaFunction, "Properties.VpcConfig") as Record<
+            string,
+            unknown
+        >;
+        expect(vpcConfig).toHaveProperty("SecurityGroupIds");
+        expect((vpcConfig.SecurityGroupIds as AwsCfInstruction[])[0]).toMatchObject({
+            Ref: computeLogicalId("vpc", "AppSecurityGroup"),
+        });
+        expect(vpcConfig).toHaveProperty("SubnetIds");
+        expect(vpcConfig.SubnetIds).toContainEqual({
+            Ref: computeLogicalId("vpc", "VPC", "PrivateSubnet1", "Subnet"),
+        });
+        expect(vpcConfig.SubnetIds).toContainEqual({
+            Ref: computeLogicalId("vpc", "VPC", "PrivateSubnet2", "Subnet"),
+        });
+    });
+    it("throws an error when using the construct twice", async () => {
+        expect.assertions(2);
+
+        try {
+            await runServerless({
+                config: Object.assign(baseConfig, {
+                    constructs: {
+                        vpc1: {
+                            type: "vpc",
+                        },
+                        vpc2: {
+                            type: "vpc",
+                        },
+                    },
+                }),
+                command: "package",
+            });
+        } catch (error) {
+            expect(error).toBeInstanceOf(ServerlessError);
+            expect(error).toHaveProperty("code", "LIFT_ONLY_ONE_VPC");
+        }
+    });
+    it("throws an error when there is an existing VPC config", async () => {
+        expect.assertions(2);
+
+        try {
+            await runServerless({
+                fixture: "vpc",
+                configExt: Object.assign(pluginConfigExt, {
+                    provider: {
+                        name: "aws",
+                        vpc: {
+                            securityGroupIds: ["sg-00000000000000000"],
+                            subnetIds: ["subnet-01234567899999999", "subnet-00000000099999999"],
+                        },
+                    },
+                }),
+                command: "package",
+            });
+        } catch (error) {
+            expect(error).toBeInstanceOf(ServerlessError);
+            expect(error).toHaveProperty("code", "LIFT_ONLY_ONE_VPC");
+        }
+    });
```

Comment:
Awesome testing the error cases. I would also add a `should create all required resources` kind-of test, like for the other construct: i.e. test that `serverless deploy` actually deploys a VPC with all the related resources in the CloudFormation template.

Reply:
Isn't it testing the CDK more than testing Lift ? I'll see what I can come up with.

Reply:
👍  have a look at the other constructs

The idea is to test what is being deployed, no matter if it's using the default values of CDK constructs or if we customize it. Lift provides the outcome, and unlike CloudFormation CDK constructs could evolve.

Reply:
Done in the latest commit